crypto/ed25519/ed25519.go

```diff
@@ -73,42 +92,57 @@ func (privKey PrivKeyEd25519) Equals(other crypto.PrivKey) bool {
 	}
 }
 
-func (privKey PrivKeyEd25519) ToCurve25519() *[32]byte {
+// ToCurve25519 takes a private key and returns its representation on
+// Curve25519. Curve25519 is birationally equivalent to Edwards25519,
+// which Ed25519 uses internally. This method is for intended use in
+// Diffie Hellman key exchanges. (i.e. X25519)
+func (privKey PrivKeyEd25519) ToCurve25519() *[PubKeyEd25519Size]byte {
 	keyCurve25519 := new([32]byte)
 	privKeyBytes := [64]byte(privKey)
 	extra25519.PrivateKeyToCurve25519(keyCurve25519, &privKeyBytes)
 	return keyCurve25519
 }
 
-// Deterministically generates new priv-key bytes from key.
+// Generate deterministically generates a new priv-key bytes from key.
+// The privkey is generated as Sha256(amino_encode({privkey, index}))
+// Note that we are using the non-standard golang definition for ed25519
+// privkeys that includes the pubkey appended to it
 func (privKey PrivKeyEd25519) Generate(index int) PrivKeyEd25519 {
-	bz, err := cdc.MarshalBinaryBare(struct {
+	bz := cdc.MustMarshalBinaryBare(struct {
 		PrivKey [64]byte
 		Index   int
 	}{privKey, index})
-	if err != nil {
-		panic(err)
-	}
 	newBytes := crypto.Sha256(bz)
 	newKey := new([64]byte)
 	copy(newKey[:32], newBytes)
+	// Caches the pubkey in the latter 32 bytes of the privkey
+	// (this is a mutative function)
```

Comment:
does this clarify / help? It does not mutate the underlying `privKey`? (What is a "mutative function" and why is it relevant in that context?) I'd propose to remove this.

Reply:
This does mutate the private key object. (I found this quite confusing tbh, it is what caused #1999)

Perhaps changing it to 
" this mutates the last 32 bytes of the privkey" would be more clear?

Reply:
It is confusing indeed. But wait, it does not mutate the private on which it is called on (which is represented by the var `privKey`, right?), it does indeed "mutate" the freshly created private-key though which is called `newKey`. Not sure if "mutative" makes sense in this context then (because of course the method "mutates" the object which it creates). Or am I getting this wrong?

Reply:
`ed25519.MakePublicKey(newKey)` alters the last 32 bytes of `newKey`